src/rules/noTautologyExpressionRule.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as tsutils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+const TAUTOLOGY_DISCOVERED_ERROR_STRING = "Expression is a tautology. Comparison is redundant.";
```

Comment:
This rule is missing the ability to do the logic mentioned in the original issue: `obj1.foo === obj1.foo`. Additionally, it could be made quite a bit more powerful by becoming an optionally-type-checked rule, so it could check cases like:

```ts
const seven = 7;

if (seven === 7) { }
```

I think the `obj1.foo === obj1.foo` case should be a built-in part of the rule when it first goes in, but it's fine to file a followup issue for the optional type checking. It'd be a shame to have the `obj1.foo` case be a breaking change after the rule is enabled... but other maintainers would have a better knowledge of what the preference is here?

`unnecessaryBindRule.ts` is a good example to go off of.

Reply:
Ehh, let's just file a followup issue. Don't want to block this from coming in.